# KNN(K Nearest Neighbours)

- Simple ML algorithm that predicts the label of a new sample based on the K closest neighbours. K any no. like n
- Lazy learner---> there is no model training, it stores just training data and tells based on that.
- Instance based learner---> it uses the actual stored data points to make predictions directly without undergoing any training it just learns and keep by seeing its neighbours as it is.
- Non parametric---> no assumptions about the shape of the data.(like bell shaped-min max etc no botheration)

## What type of ML algorithm it is?
- supervised
- can be used for both regression and classification
- because the output is decided based on the neighbours: if we check:
  - majority vote---> classification
  - average of the neighbours---> regression

## Where is KNN used?
- data is small or medium sized where storing the data is not expensive as it does not use model instead stores the data as it is. around 500 rows
- whenever decision boundary is irregular or non linear ( logistic regression fails here as logistic is smooth curve and KNN can adapt locally)

## How does KNN actually work?

for any new sample/test data:

Step 1: Calculate the Euclidean distance between the new sample point and ALL training samples.

Step 2: Arrange the distance and pick the K nearest neighbours to that sample point.K should be an odd value:3,5,7
Because if 4 and 2 says one and other 2 says something we cant come to conclusion

Step 3:
- Classification---> majority vote
- Regression---> average of neighbour values

Step 4:
- Assign the predicted output.
- No "learn" function.It waits untill prediction to take further action.





## Implementation

# Step 1: Problem statement

In [ ]:
#Predict whether the tumor is benign or malignant

# Step 2: Load Dataset

In [ ]:
# load the dataset from sklearn
# convert df
from sklearn.datasets import load_breast_cancer
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

data= load_breast_cancer(as_frame=True)
df=data.frame # to get even target
df.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


# Step 3: Preprocessing

In [ ]:
# separate the features and the target
x= df.drop('target',axis=1) # all columns except the target
y= df['target']

In [ ]:
# scaling using standard scaler can be done
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
x_scaled=scaler.fit_transform(x) # scaled version of features

# Step 4: Train-Test Split

In [ ]:
##                 x(features)                 y(output)
##
##Training(70%)-->   x_train                   y_train
##Testing(30%)-->    x_test                    y_test

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42) # order of variables like this important

In [ ]:
# shape
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((398, 30), (171, 30), (398,), (171,))

# Step 5: Train the model

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn= KNeighborsClassifier(n_neighbors=5) #k=3,5,7,so on 7 no need as might be widely spread and then wont be nearest

knn.fit(x_train,y_train) # no as such training it its just rough training calculating distance and fininding neighbours and testing for it

KNeighborsClassifier()

# Step 6: Evaluate the model

In [ ]:
# prediction # based on that we predict and see all evaluations how it has come based on that we can change
y_pred= knn.predict(x_test)

In [ ]:
# calculate accuracy,precision,recall,f1 score,roc auc score
from sklearn.metrics import roc_auc_score
yprob_a=knn.predict_proba(x_test)[:,1]
roc_auc=roc_auc_score(y_test,yprob_a)

In [ ]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score
# given as list as works better
metric_df=pd.DataFrame({
    'Model':['KNN'],
    'Accuracy':[accuracy_score(y_test,y_pred)],
    'Precision':[precision_score(y_test,y_pred)],
    'Recall':[recall_score(y_test,y_pred)],
    'F1 score':[f1_score(y_test,y_pred)],
    'ROC AUC':[roc_auc]
})

metric_df

,Model,Accuracy,Precision,Recall,F1 score,ROC AUC
0,KNN,0.959064,0.946903,0.990741,0.968326,0.995297


In [ ]:
metric_df.round(4)

,Model,Accuracy,Precision,Recall,F1 score,ROC AUC
0,KNN,0.9591,0.9469,0.9907,0.9683,0.9953


# Step 7: Interpret results

Accuracy - model correctly predicted the outcome for about 95.9% of the data in the test set

Precision -  model predicts a class as positive, it is correct about 94.7% of the time

Recall - model successfully finds 99.1% of all the actual positive cases

F1 Score - balance between model's precision and recall (96%)

ROC AUC - model has a very good overall performance and is highly capable of correctly ranking positive and negative observations

there are other algorithms also

# Hyperparameter Tuning

## What are Hyperparameters?

- Model parameters?
  - Learned from data during training
  - Change automatically when you call 'fit()'
  - Eg:
    - Logistic regression: coefficients w1,w1,intercept b
    - Neural networks: weights for all layers

- Hyper Parameters
  - **set by you,before training**
  - they control how the algorithm learns or behaves,ground rules for model to learn
  - they are not learned from the data,the algorithm treats them as fixed.
  - Example:
    - KNN: no.of neighbours was fixed(5) the algorithm shd only look for closest neighbours and not anybody else, and distance by defalut we said eucledian is a hyperparameter for euclidean there are other manhattan

## why do we need hyperparameters?

- different hyperparameters values can give different performance.(values)
- wrong hyperparameters---> underfitting or overfitting
  - Underfitting: Model generalises info from neighbors and does not calculate

## what are the methods of hyperparameter tuning(in ML terms)

- Underfitting:
  - model is too simple
  - Ignores local patterns---> leads to high bias and low variance
- Overfitting:
  - model is too complex and learns the training data's noise and performs well on the training data but poorly on test data.
  - model memorises the data and just gives it as it is ---> very low bias,high varaince

When is a model considered good?
- when the training data patterns are correctly understood by the model
- when it generalizes well to new,unseen data
- not memorize the training data(overfitting) or fail to capture relevant patterns(underfitting)

## How to implement it?

will search and tell if model good considering hyperparameters
1. Manual Search- not suggested as not systematic- cant do trail and error method manually
2. Grid Search(really exhaustive)- `GridSearchCV`

- values for each hyperparameter it will check and convert this to a grid
- the algorithm will test every single combination using cross- validation.

3. Random Search- `RandomizedSearchCV`

- you can define a range for each hyperparameter
- randomly sample combinations and evaluate them
- more effective

more like bayesian optimization

doing for knn the above dataset

In [ ]:
# and splitting dataset:into x and y,we split using the train-test split
# x--> scaled(important if in KNN)

# x_train,x_test,y_train and y_test

In [ ]:
# GridSearchCV

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

#define the parameter grid using weights and defining
param_grid={
    # 'hyperparameter': ['value1','value2','value3']
    'n_neighbors': [3,5,7,9,11], # we are giving specific values not range of values # a range of values in Randomized search
    'weights': ['uniform','distance'],
    'metric':['minkowski'], # error calculation
    'p':[1,2] # p=1--> manhattan,p=2---> eucledian
}
# leaf size

In [ ]:
grid_search=GridSearchCV(
    estimator=knn,
    param_grid=param_grid,
    scoring='accuracy', # how well optimized
    cv=5 # 5 fold cross validation
)

In [ ]:
# run the grid
grid_search.fit(x_train,y_train)

# best parameters
grid_search.best_params_

{'metric': 'minkowski', 'n_neighbors': 5, 'p': 1, 'weights': 'uniform'}

In [ ]:
# cv accuracy for this combination
grid_search.best_score_

np.float64(0.9245886075949368)

In [ ]:
from sklearn.metrics import accuracy_score,classification_report

best_knn=grid_search.best_estimator_ # knn changed to grid search's best_estimator_ in that only training done so now only prediction with new knn

y_pred_best=best_knn.predict(x_test)

accuracy_best=accuracy_score(y_test,y_pred_best)
accuracy_best

0.9590643274853801

In [ ]:
metric_df.loc[1]=[
    'Grid search',
    accuracy_score(y_test,y_pred_best),
    precision_score(y_test,y_pred_best),
    recall_score(y_test,y_pred_best),
    f1_score(y_test,y_pred_best),
    roc_auc_score(y_test,y_pred_best)]


metric_df

# increased
# this is called as hyperparameter tuning by setting

,Model,Accuracy,Precision,Recall,F1 score,ROC AUC
0,KNN,0.959064,0.946903,0.990741,0.968326,0.995297
1,Grid search,0.959064,0.954955,0.981481,0.968037,0.951058
